src/librustc_parse/parser/attr.rs

```diff
@@ -273,17 +273,27 @@ impl<'a> Parser<'a> {
 
         let lo = self.token.span;
         let path = self.parse_path(PathStyle::Mod)?;
-        let kind = self.parse_meta_item_kind()?;
+        let kind = self.parse_meta_item_kind(&path)?;
         let span = lo.to(self.prev_token.span);
         Ok(ast::MetaItem { path, kind, span })
     }
 
-    crate fn parse_meta_item_kind(&mut self) -> PResult<'a, ast::MetaItemKind> {
+    crate fn parse_meta_item_kind(&mut self, path: &ast::Path) -> PResult<'a, ast::MetaItemKind> {
         Ok(if self.eat(&token::Eq) {
             ast::MetaItemKind::NameValue(self.parse_unsuffixed_lit()?)
         } else if self.check(&token::OpenDelim(token::Paren)) {
-            // Matches `meta_seq = ( COMMASEP(meta_item_inner) )`.
-            let (list, _) = self.parse_paren_comma_seq(|p| p.parse_meta_item_inner())?;
+            let (list, _) = if path == &rustc_span::sym::version {
```

Comment:
This is incorrect, `version(...)` shouldn't allow semver expression like `1.2.3` in arbitrary attributes, only in `cfg`.

Reply:
As a result, `MetaItemKind::List` cannot be used for `version(1.2.3)` in `eval_condition` because it's not a meta-item.

You also cannot use `span_to_snippet` for obtaining the text representation of `1.2.3` in general because spans are not reliable.

I'd recommend implementing `version("1.2.3")` with a string literal in an initial PR (which *is* a meta-item and is already stringified), and then think about parsing and stringifying `1.2.3` instead.

Reply:
"1.2.3" was already done in the previous commits.

Reply:
The RFC uses `version(1.27.0)` as example IIRC.

Reply:
Ok, I'm going to go back to the original string literal version